tinygrad/renderer/llvmir.py

```diff
@@ -145,7 +146,7 @@ def store_op(): bb[-1].store(element, bb[-1].gep(lvars[vin[0]], [lvars[vin[1]]],
         with bb[-1].if_then(bb[-1].trunc(lvars[vin[3]], ir.IntType(1))): store_op()
       else: store_op()
     if uop == UOps.ALU:
-      lvars[u] = cast(bb, code_for_op[args](bb[-1], *[cast(bb, lvars[x], x.dtype, dtypes.float) for x in vin]), dtypes.float, dtype)
+      lvars[u] = cast(bb, code_for_op[args](bb[-1], *[cast(bb, lvars[x], x.dtype, dtypes.float) for x in vin]), dtypes.float, dtype) if not (args == UnaryOps.NEG and dtype == dtypes.bool) else bb[-1].xor(lvars[vin[0]], ir.Constant(ir.IntType(1), 1))
```

Comment:
This logic should probably be in code_for_op

Reply:
ok two options:
allow code_for_op to take a dtype https://github.com/tinygrad/tinygrad/compare/master...Qazalin:tinygrad:code-for-op-with-dtype
Implement a new UnaryOps.NOT https://github.com/tinygrad/tinygrad/compare/master...Qazalin:tinygrad:UnaryOps-NOT
[similar to torch's logical_not() ](https://pytorch.org/docs/stable/generated/torch.logical_not.html)

Reply:
So definitely not on the new op.

If code_for_op takes a dtype (which I'm largely fine with), what does that dtype apply to? Inputs? Outputs? Both?

Reply:
both, inputs always get upcasted to the final output's dtype